spec/Section 3 -- Type System.md

```diff
@@ -1679,6 +1751,12 @@ defined.
    the original Input Object.
 4. Any non-repeatable directives provided must not already apply to the original
    Input Object type.
+5. The `@oneOf` directive must not be provided by an Input Object type
+   extension.
```

Comment:
Why is this the case?  extend is useful for tooling to "create types" in a peicemeal fashion.  Not everyone needs to know all the details but it needs to be validated at the end at actual schema creation time

```
#fileA.graphqls
input Foo {
   a : String
}
```

```
#fileB.graphqls
extend input Foo @oneOf {
   b : String
}
```

The following once combined is valid surely?

Reply:
Twofold: it’s because a ”OneOf Input Object” may be constructed in a different way to traditional input objects (might use a different class or similar), and because if this was not the case then when the directive was added you’d have to go back and validate all previously added fields since they have different “potential to be invalid” text.

Reply:
Hmm kinda feels like this is project an implementation into the spec situation.  

> it’s because a ”OneOf Input Object” may be constructed in a different way to traditional input objects (might use a different class or similar)

I would argue that the spec should not care how its implemented - sure it could be a new OneOfInputType class at implementation time (I toyed with this in graphql-java but rejected it as too much impact) but the spec shouldn't care.

> When the directive was added you’d have to go back and validate all previously added fields 

Again an implementation detail - not sure that the spec should care.

Even merging them seems straightforward since they are a non repeating directive

```
#fileA.graphqls
input Foo {
   a : String!
}
```

```
#fileB.graphqls
extend input Foo @oneOf {
   b : String
}
```

```
#fileC.graphqls
extend input Foo @oneOf {
   c : String
}
```

And implementation can merge these easily enough.

Right now the spec says about input types

```
extend input [Name] [Directives]opt [InputFieldsDefinition]
extend input [Name] [Directives]opt 
```

in other words it allows you to put extra directives on the input types

I just feel like this is an unnecessary restriction

Reply:
> I would argue that the spec should not care how its implemented - sure it could be a new OneOfInputType class at implementation time (I toyed with this in graphql-java but rejected it as too much impact) but the spec shouldn't care.

I agree, I'd argue that implementing Input Objects and OneOf Input Objects as the _same_ class in the implementation is an implementation detail that the space doesn't need to worry about 😉 

At one point, oneOf was going to be its own top-level type, like `oneof Foo { a: Int, b: Float }` rather than `input Foo @oneof { a: Int, b: Float }`. Would you feel the same if we had chosen the alternative syntax?

To me, a OneOf Input Object is a distinct type in GraphQL that happens to share a lot with Input Objects, but still has it's own distinct behaviors in many ways. As such, changing an Input Object into a OneOf Input Object would be like changing an Input Object into an Object for me - and that's not something I think the SDL should allow.

This is one of the arguments _against_ using the directive syntactical representation in my opinion, and encouraging using a keyword instead - to make the separation clearer.

> Again an implementation detail - not sure that the spec should care.

This second point is not an implementation detail, it's a spec detail - specifically it relates to lines 1724-1726 in `spec/Section 3 -- Type System.md` - that's the validation rules for the _definition_ of an `input` - if it's defined as `input Foo @oneOf` then additional rules apply. Then similar rules are applied on lines 1756-1759 for fields added via an _input type extension_. If we allow the directive to be added during an _input type extension_ then the previously performed rules in the _input type definition_ would now be invalidated and we'd need to re-check them; I can't think of a place where we've done this kind of back reference and re-validation in a similar part of the spec.

Reply:
Along the same thoughts that @benjie is pointing out, I personally would say a different syntactical representation in the SDL would make it appear as a "distinct type", whereas since it is a directive, it is simply additional validation rules on top of the existing input object type.  Keep in mind that directives are described as such:

> A GraphQL schema describes directives which are used to annotate various parts of a GraphQL document as an indicator that they should be evaluated differently by a validator, executor, or client tool such as a code generator.

I think this definition fits `@oneOf`.  This PR also describes oneOf input objects as such:

> OneOf Input Objects are a special variant of Input Objects

So the way this PR is written, I see `@oneOf` as a predefined directive, just like `@skip` or `@include` or `@deprecated`, that converts the input object it is decorating into the 'special variant'.  As such, I would personally agree with @bbakerman in that the directive should be able to be applied anywhere it has always been, and hence the extra validation checks applied if the schema requires it after being merged.

Reply:
> OneOf Input Objects are a special variant of Input Objects

Indeed, they are a special variant. In my opinion you wouldn't take a broad thing and then turn it into a special variant at a later stage - it needs to be a special variant from the start. But this is my opinion, I'm happy to test it against the broader working group.

---

Imagine for a second that we didn't have input objects. You could imagine taking an object type and tagging it `@input`:

```graphql
type UserInput @input {
  id: ID!
  name: String!
  bio: String
}
```

This syntactic change could indicate:

1. that the type cannot have arguments on any fields
2. that the unwrapped type of each of the fields can only use scalars, enums, or other `@input` types
3. that the type cannot be used as the type of any field, other than a field on an `@input` type
4. that the type can now be used as the type of arguments

This is essentially the input object type at this point. But all we've done as annotated the existing object type with different validation logic and execution behaviors. Is the input object type really not distinct from the object type? Would turning an object type into an `@input` object type be a reasonable thing to do later down in the SDL? I, personally, don't think so. I think doing so would be extremely confusing for people reading the SDL - where something suddenly changes from one type of thing to another.

This, again, is one of the big issues that I have with using a directive to represent this distinct type. (I'm also in favour of using the directive because it's the smallest change, and allows it to be compatible with existing clients.) This constraint: that you can't add the directive in an extension, is my compromise - to allow us to leverage the benefits of using a directive without the footguns in terms of changing one thing into another at a later stage.